mllib/src/main/scala/org/apache/spark/ml/tree/Split.scala

```diff
@@ -189,3 +189,76 @@ class ContinuousSplit private[ml] (override val featureIndex: Int, val threshold
     OldSplit(featureIndex, threshold, OldFeatureType.Continuous, List.empty[Double])
   }
 }
+
+/**
+ * Split with pointers to the left and right child nodes in decision tree.
+ * Used to update the node index cache.
+ * @param split Input split
+ */
+class SplitWithChildNodeInfo private[ml] (split: Split) extends Split {
+  private var _leftChildIndex : Option[Int] = None
+  private var _rightChildIndex : Option[Int] = None
+
+  /**
+   * Create split with pointers to left and right child nodes
+   * @param split Input split
+   * @param leftChildIndex Index of left child node
+   * @param rightChildIndex Index of right child node
+   */
+  def this(split: Split, leftChildIndex: Int, rightChildIndex: Int) = {
+    this(split)
+    setLeftChildNodeId(leftChildIndex)
+    setRightChildNodeId(rightChildIndex)
+  }
+
+  /** Index of feature which this split tests */
+  override def featureIndex: Int = split.featureIndex
+
+  /**
+   * Return true (split to left) or false (split to right).
+   * @param features  Vector of features (original values, not binned).
+   */
+  override private[ml] def shouldGoLeft(features: Vector): Boolean = {
+    split.shouldGoLeft(features)
+  }
+
+  /**
+   * Return true (split to left) or false (split to right).
+   * @param binnedFeature Binned feature value.
+   * @param splits All splits for the given feature.
+   */
+  override private[tree] def shouldGoLeft(binnedFeature: Int, splits: Array[Split]): Boolean = {
+    split.shouldGoLeft(binnedFeature, splits)
+  }
+
+  /** Convert to old Split format */
+  override private[tree] def toOld: OldSplit = {
+    split.toOld
+  }
+
+  /** Set index of left child node in decision tree */
+  def setLeftChildNodeId(nodeId: Int) = {
+    _leftChildIndex = Some(nodeId)
+  }
+
+  /** Set index of right child node in decision tree */
+  def setRightChildNodeId(nodeId: Int) = {
+    _rightChildIndex = Some(nodeId)
+  }
+
+  /** Get index of left child node in decision tree */
+  def getLeftChildNodeId : Int = {
+    _leftChildIndex match {
+      case Some(id) => id
+      case _ => throw new RuntimeException("Left child node id is not defined")
```

Comment:
Why are we resolving the Option here?  The caller will have to be checking depth or isLeaf before calling?  I guess either way the caller has to handle it.